File: python/open_guardrail/guards/mcp_tool_safety.py
"""Validates MCP (Model Context Protocol) tool calls for safety."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_DEFAULT_DENIED_TOOLS = [
    "execute_command",
    "run_shell",
    "write_file",
    "delete_file",
    "send_request",
]

_MCP_PATTERNS = [
    re.compile(r"mcp://", re.IGNORECASE),
    re.compile(r"\bserver:\s*", re.IGNORECASE),
    re.compile(r"\btool:\s*", re.IGNORECASE),
    re.compile(r"\bmcp_call\b", re.IGNORECASE),
    re.compile(r"\buse_mcp_tool\b", re.IGNORECASE),
]

_SERVER_PATTERN = re.compile(r"(?:mcp://|server:\s*)(\S+)", re.IGNORECASE)
_TOOL_PATTERN = re.compile(r"(?:tool:\s*|mcp_call\s+|use_mcp_tool\s+)(\S+)", re.IGNORECASE)


class _McpToolSafetyGuard:
    def __init__(
        self,
        *,
        action: str = "block",
        allowed_servers: Optional[List[str]] = None,
        denied_tools: Optional[List[str]] = None,
    ) -> None:
        self.name = "mcp-tool-safety"
        self.action = action
        self.allowed_servers = allowed_servers
        self.denied_tools = denied_tools if denied_tools is not None else _DEFAULT_DENIED_TOOLS

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        issues: List[str] = []

        has_mcp = any(p.search(text) for p in _MCP_PATTERNS)
        if not has_mcp:
            elapsed = (time.perf_counter() - start) * 1000
            return GuardResult(
                guard_name="mcp-tool-safety",
                passed=True,
                action="allow",
                message=None,
                latency_ms=round(elapsed, 2),
                details=None,
            )

        server_match = _SERVER_PATTERN.search(text)
        if self.allowed_servers and server_match:
            server = server_match.group(1)
            if server not in self.allowed_servers:
                issues.append(f"Unauthorized MCP server: {server}")

        tool_match = _TOOL_PATTERN.search(text)
        if tool_match:
            tool = tool_match.group(1)
            if tool in self.denied_tools:
                issues.append(f"Denied MCP tool: {tool}")

        for d in self.denied_tools:
            if d in text and not any(d in i for i in issues):
                issues.append(f"Denied tool reference found: {d}")

        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="mcp-tool-safety",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message="; ".join(issues) if triggered else None,
            latency_ms=round(elapsed, 2),
            details={
                "issues": issues,
                "server": server_match.group(1) if server_match else None,
                "tool": tool_match.group(1) if tool_match else None,
            } if triggered else None,
        )


def mcp_tool_safety(
    *,
    action: str = "block",
    allowed_servers: Optional[List[str]] = None,
    denied_tools: Optional[List[str]] = None,
) -> _McpToolSafetyGuard:
    return _McpToolSafetyGuard(
        action=action,
        allowed_servers=allowed_servers,
        denied_tools=denied_tools,
    )

File: packages/guards/src/mcp-tool-safety.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface McpToolSafetyOptions {
  action: 'block' | 'warn';
  allowedServers?: string[];
  deniedTools?: string[];
}

const DEFAULT_DENIED_TOOLS: string[] = [
  'execute_command',
  'run_shell',
  'write_file',
  'delete_file',
  'send_request',
];

const MCP_PATTERNS: RegExp[] = [
  /mcp:\/\//i,
  /\bserver:\s*/i,
  /\btool:\s*/i,
  /\bmcp_call\b/i,
  /\buse_mcp_tool\b/i,
];

const SERVER_PATTERN = /(?:mcp:\/\/|server:\s*)(\S+)/i;
const TOOL_PATTERN = /(?:tool:\s*|mcp_call\s+|use_mcp_tool\s+)(\S+)/i;

export function mcpToolSafety(options: McpToolSafetyOptions): Guard {
  const denied = options.deniedTools ?? DEFAULT_DENIED_TOOLS;
  const allowed = options.allowedServers;

  return {
    name: 'mcp-tool-safety',
    version: '0.1.0',
    description:
      'Validates MCP (Model Context Protocol) tool calls for safety',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      const hasMcp = MCP_PATTERNS.some((p) => p.test(text));
      if (!hasMcp) {
        return {
          guardName: 'mcp-tool-safety',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      const serverMatch = text.match(SERVER_PATTERN);
      if (allowed && serverMatch) {
        const server = serverMatch[1];
        if (!allowed.includes(server)) {
          issues.push(`Unauthorized MCP server: ${server}`);
        }
      }

      const toolMatch = text.match(TOOL_PATTERN);
      if (toolMatch) {
        const tool = toolMatch[1];
        if (denied.includes(tool)) {
          issues.push(`Denied MCP tool: ${tool}`);
        }
      }

      for (const d of denied) {
        if (text.includes(d) && !issues.some((i) => i.includes(d))) {
          issues.push(`Denied tool reference found: ${d}`);
        }
      }

      const triggered = issues.length > 0;
      return {
        guardName: 'mcp-tool-safety',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? {
              issues,
              server: serverMatch?.[1],
              tool: toolMatch?.[1],
            }
          : undefined,
      };
    },
  };
}
